src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -496,72 +517,92 @@ func (c controller) postInstallConfigs(ctx context.Context) (error, map[string]i
 	return nil, nil
 }
 
-func (c controller) waitForOLMOperators(ctx context.Context) error {
+func (c *controller) isStageTimedOut(stage models.FinalizingStage) func() bool {
+	return func() bool {
+		cluster, err := c.ic.GetCluster(context.TODO(), false)
+		if err != nil {
+			c.log.WithError(err).Error("failed to get cluster")
+			return false
+		}
+		return cluster != nil && cluster.Progress != nil && cluster.Progress.FinalizingStage == stage && cluster.Progress.FinalizingStageTimedOut
+	}
+}
+
+func (c *controller) waitForOLMOperators(ctx context.Context) error {
 	var operators []models.MonitoredOperator
-	var err error
+	var err, retryErr error
+	var errs []error
 
 	// In case the timeout occur, we have to update the pending OLM operators to failed state,
```

Comment:
still relevant? I don't understand a little bit here too, if all the timeouts are part of service so why controller should set failed on operator for timeout? I think that if we move all the timeout decisions to service we should handle operators timeout failure in service too.

Reply:
For the same reason the controller notifies success.
And the service cannot decide on success unless the controller notifies (with same API call).

Reply:
But service on other hand decides that it got timeout so it should set timeout per operator and not to wait for controller to do it. It sort of service decides that timeout occurred and only then controller sends failed to operator cause service decided that operator failed. I am not sure i understand why in this case we need controller to send failed state. It can be next ticket but i really think if we already moving failure decision to service we should manage full failure flow from service and not to wait for controller to set it cause controller doesn't decide anything here, i actually would expect that service should decide success too by looking on operator status